### PR TITLE
HSC-17: Support for Bahmni Forms 2.0.

### PR DESCRIPTION
**Pull request by chinzou**
https://mekomsolutions.atlassian.net/browse/HSC-17

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -99,6 +100,14 @@
...
       <url>http://repo1.maven.org/maven2</url>
     </repository>
+      <repository>
+        <id>repo.mybahmni.org</id>
+        <name>bahmni-artifactory-snapshots</name>
+        <url>http://repo.mybahmni.org.s3.amazonaws.com/artifactory/snapshot</url>
+        <snapshots>
+          <updatePolicy>always</updatePolicy>
+        </snapshots>
+      </repository>
```
Comment (rbuisson) [inline pom.xml]: @chinzou , I think you should add the release repo too.

File: pom.xml [hunk 2]
```diff
@@ -99,6 +100,27 @@
...
+<!--          <updatePolicy>always</updatePolicy>-->
+<!--        </snapshots>-->
+<!--      </repository>-->
+<!--    <repository>-->
+<!--      <id>repo.mybahmni.org-release</id>-->
+<!--      <name>bahmni-artifactory-release</name>-->
+<!--      <url>http://repo.mybahmni.org.s3.amazonaws.com/artifactory/release</url>-->
+<!--    </repository>-->
+
+<!--    this repository is temporary due to Bahmni IE Apps not available on Bahmni repos yet -->
```
Comment (mks-d) [inline pom.xml]: Can you put the explanation _above_ the snippet that is commented out?
Is there any harm to **not** comment this out?

File: api/src/test/java/org/openmrs/module/initializer/api/form/FormLoaderIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	}
+	
+	@Test
+	public void load_shouldLoadFormAccordingToJsonFile() {
+		
+		// Replay
+		formLoader.load();
+		
+		// Verify
+		Assert.assertNotNull(formService.getForm("test"));
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/form/FormLoaderIntegrationTest.java]: We need to assert that the form object is as expected given the test Json definition file.

File: api/src/test/java/org/openmrs/module/initializer/api/form/FormLoaderIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	}
+	
+	@Test
+	public void load_shouldSaveFormTranslation() {
+		
+		// Replay
+		formLoader.load();
+		
+		// Verify
+		Assert.assertNotNull(bahmniFormTranslationService.getFormTranslations("test", "1", null));
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/form/FormLoaderIntegrationTest.java]: We need to assert that the translations are as expected given the test Json definition file.

File: api/src/test/java/org/openmrs/module/initializer/api/form/FormLoaderIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	}
+	
+	@Test
+	public void load_shouldCreateFormJsonFile() {
+		
+		// Replay
+		formLoader.load();
+		
+		// Verify
+		Assert.assertTrue((new File(formFolderPath + "test_1.json")).exists());
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/form/FormLoaderIntegrationTest.java]: Can't any relevant asserts be made on the **content** of this .json file?

File: api/src/test/java/org/openmrs/module/initializer/api/form/FormLoaderIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	}
+	
+	@Test
+	public void load_shouldCreateFormTranslationJsonFile() {
+		
+		// Replay
+		formLoader.load();
+		
+		// Verify
+		Assert.assertTrue((new File(formTranslationPath + "test_1.json")).exists());
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/form/FormLoaderIntegrationTest.java]: Can't any relevant assert be made on the **content** of this .json file?

File: pom.xml (hunk 2 above)
Comment (chinzou) [reply pom.xml]: yes, there is no harm to comment this out :)

File: api/pom.xml [hunk 7]
```diff
@@ -82,6 +82,11 @@
       <version>${exti18nVersion}</version>
       <scope>test</scope>
     </dependency>
+		<dependency>
+			<groupId>org.bahmni.module</groupId>
+			<artifactId>bahmni.ie.apps-api</artifactId>
+			<version>${bahmniIeAppsVersion}</version>
+		</dependency>
```
Comment (rbuisson) [inline api/pom.xml]: @chinzou , I believe we should set the dependency as `<scope>provided</scope>` in order to make sure we use the version provided at runtime.

File: api/src/main/java/org/openmrs/module/initializer/Domain.java [hunk 8]
```diff
@@ -24,7 +24,8 @@
 	ORDER_TYPES(20, "ordertypes"),
 	APPOINTMENTS_SPECIALITIES(21, "appointmentsspecialities"),
 	APPOINTMENTS_SERVICES_DEFINITIONS(22, "appointmentsservicesdefinitions"),
-	DATAFILTER_MAPPINGS(23, "datafiltermappings");
+	DATAFILTER_MAPPINGS(23, "datafiltermappings"),
+	FORM(24, "form");
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/Domain.java]:
```java
BAHMNI_FORMS(24, "bahmniforms");
```

File: api/src/main/java/org/openmrs/module/initializer/api/form/FormLoader.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+@OpenmrsProfile(modules = { "bahmni.ie.apps:*" })
+@Component
+public class FormLoader extends BaseLoader {
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/form/FormLoader.java]:
```
public class BahmniFormsLoader extends BaseLoader {
```

File: api/src/main/java/org/openmrs/module/initializer/Domain.java [hunk 10]
```diff
@@ -24,7 +24,8 @@
 	ORDER_TYPES(20, "ordertypes"),
 	APPOINTMENTS_SPECIALITIES(21, "appointmentsspecialities"),
 	APPOINTMENTS_SERVICES_DEFINITIONS(22, "appointmentsservicesdefinitions"),
-	DATAFILTER_MAPPINGS(23, "datafiltermappings");
+	DATAFILTER_MAPPINGS(23, "datafiltermappings"),
+	BAHMNI_FORMS(24, "bahmniforms");
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/Domain.java]: Looks like that's too low on the list, what other metadata do Bahmni forms need to be loaded?

File: api/src/test/resources/org/openmrs/module/initializer/include/gp_bahmni_forms.xml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+<config>
+    <globalProperties>
+        <globalProperty>
+            <property>bahmni.forms.directory</property>
+            <defaultValue>/src/test/resources/</defaultValue>
+            <description>
```
Comment (mks-d) [inline api/src/test/resources/org/openmrs/module/initializer/include/gp_bahmni_forms.xml]:
```xml
<description>
  File system path for saving Bahmni forms.
</description>
```

File: api/src/test/resources/org/openmrs/module/initializer/include/gp_bahmni_forms.xml [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+                Filesystem path for saving the bahmni forms
+            </description>
+        </globalProperty>
+
+        <globalProperty>
+            <property>bahmni.formTranslations.directory</property>
+            <defaultValue>src/test/resources/</defaultValue>
+            <description>
+                Filesystem path for saving the bahmni forms translations
+            </description>
```
Comment (mks-d) [inline api/src/test/resources/org/openmrs/module/initializer/include/gp_bahmni_forms.xml]:
```xml
<description>
  File system path for saving Bahmni forms translations.
</description>
```

File: api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+				is = new FileInputStream(file);
+				Map jsonFile = new ObjectMapper().readValue(is, Map.class);
+				Map formJson = (Map) jsonFile.get("formJson");
+				String formName = (String) formJson.get("name");
+				List<Map> resources = (List<Map>) formJson.get("resources");
+				String resourceValue = (String) resources.get(0).get("value");
+				
+				// Save OpenMRS form
+				Form form = new Form();
+				form.setVersion("1");
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java]: What's that version? Can you explain or point to resources to read?

Comment (chinzou) [reply api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java]: As in Form Builder, the version is explicitly set to "1" [here](https://github.com/Bahmni/implementer-interface/blob/d396bb712e96e594571d0e23077b2f6185cf8c9f/src/form-builder/components/FormBuilder.jsx#L68)

File: api/src/main/java/org/openmrs/module/initializer/Domain.java [hunk 14]
```diff
@@ -26,7 +26,8 @@
 	DATAFILTER_MAPPINGS(22, "datafiltermappings"),
 	METADATA_SETS(23, "metadatasets"),
 	METADATA_SET_MEMBERS(24, "metadatasetmembers"),
-	METADATA_TERM_MAPPINGS(25, "metadatatermmappings");
+	METADATA_TERM_MAPPINGS(25, "metadatatermmappings"),
+	BAHMNI_FORMS(26, "bahmniforms");
```
Comment (rbuisson) [inline api/src/main/java/org/openmrs/module/initializer/Domain.java:30]: @chinzou , as Dimitri said, we must asses if that makes sense to load this so late in the list.
I think Bahmni Forms do not depend on much other metadata, therefore they can probably be loaded quite eaerly.

File: api/src/test/resources/org/openmrs/module/initializer/include/gp_bahmni_forms.xml [hunk 15]
```diff
@@ -0,0 +1,19 @@
+<config>
```
Comment (rbuisson) [inline api/src/test/resources/org/openmrs/module/initializer/include/gp_bahmni_forms.xml:1]: @chinzou , is this file really needed? I see in the `setup()` method of `BahmniFormsLoaderIntegrationTest` that the GP are saved there as well. Remove if not needed.

File: api/src/test/java/org/openmrs/module/initializer/api/form/BahmniFormsLoaderIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	
+	@After
+	public void clean() throws IOException {
+		// Delete created form files
+		
+		FileUtils.deleteDirectory(new File(formFolderPath));
+		FileUtils.deleteDirectory(new File(formTranslationPath));
+	}
+
```
Comment (rbuisson) [inline api/src/test/java/org/openmrs/module/initializer/api/form/BahmniFormsLoaderIntegrationTest.java:59]: @chinzou , make sure to test the new behavior:
- should update an already published form.
- should not create a new record (in the `form`)

File: api/src/main/java/org/openmrs/module/initializer/Domain.java (hunk 14 above)
Comment (chinzou) [reply api/src/main/java/org/openmrs/module/initializer/Domain.java:30]: @rbuisson , yes.
As we discussed I'll move it to order **10**

File: api/src/test/resources/org/openmrs/module/initializer/include/gp_bahmni_forms.xml (hunk 15 above)
Comment (chinzou) [reply api/src/test/resources/org/openmrs/module/initializer/include/gp_bahmni_forms.xml:1]: Yes. since GP are set in `setup()`, we don't need the file anymore

File: api/src/test/java/org/openmrs/module/initializer/api/form/BahmniFormsLoaderIntegrationTest.java [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+		// Setup
+		bahmniFormsLoader.load();
+		
+		// Replay
+		bahmniFormsLoader.load();
+		
+		//Verify
+		Assert.assertNull(formService.getForm("test", "2"));
+		Assert.assertEquals(1,
+		    formService.getAllForms().stream().filter(form -> form.getName().equals("test")).toArray().length);
```
Comment (rbuisson) [inline api/src/test/java/org/openmrs/module/initializer/api/form/BahmniFormsLoaderIntegrationTest.java:85]: @chinzou , could this be 1 line?

**Comment by rbuisson** [comment]
@mks-d , would you be able to have a look at this PR?

File: api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+				List<Map> resources = (List<Map>) formJson.get("resources");
+				String resourceValue = (String) resources.get(0).get("value");
+				
+				// check if form exist
+				Form existingForm = formService.getForm(formName);
+				Form form;
+				
+				if (existingForm != null) {
+					form = existingForm;
+					form.setPublished(false);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java:73]: I think this line should go.

File: api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+				} else {
+					form = new Form();
+					form.setVersion("1");
+				}
+				
+				// Save OpenMRS form
+				form.setPublished(false);
+				form.setName(formName);
+				
+				Form savedForm = formService.saveForm(form);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java:83]: reuse `form` no?
```java
form = formService.saveForm(form);
```

File: api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+				
+				// Save Bahmni form resource
+				BahmniForm bahmniForm = new BahmniForm();
+				bahmniForm.setName(savedForm.getName());
+				bahmniForm.setVersion(savedForm.getVersion());
+				bahmniForm.setUuid(savedForm.getUuid());
+				bahmniForm.setPublished(false);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java:90]: Why not
```java
bahmniForm.setPublished(savedForm.getPublished());
```
?
That's for code consistency of course, since they seem to be always unpublished.

File: api/src/test/java/org/openmrs/module/initializer/DomainBaseModuleContextSensitiveTest.java [hunk 21]
```diff
@@ -73,6 +73,11 @@ public DomainBaseModuleContextSensitiveTest() {
 			mod.setFile(new File(""));
 			ModuleFactory.getStartedModulesMap().put(mod.getModuleId(), mod);
 		}
+		{
+			Module mod = new Module("", "bahmni.ie.apps", "", "", "", "");
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/DomainBaseModuleContextSensitiveTest.java:77]: Give it a version, just for readability. But there can be cases where this actually matters, even in tests. We just got into that with ID Gen's `AutoGenerationOption`.

File: api/src/test/java/org/openmrs/module/initializer/api/form/BahmniFormsLoaderIntegrationTest.java [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class BahmniFormsLoaderIntegrationTest extends DomainBaseModuleContextSensitiveTest {
+	
+	private String formFolderPath = "src/test/resources/forms/";
+	
+	private String formTranslationPath = "src/test/resources/formsTranslation/";
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/form/BahmniFormsLoaderIntegrationTest.java:28]: If anything this path should be named `"src/test/resources/formsTranslations"` or `"src/test/resources/formTranslations"`. But why not do a subfolder: `"src/test/resources/forms/translations"`?

File: api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java (hunk 20 above)
Comment (chinzou) [reply api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java:90]: yes, sure.

File: api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java (hunk 19 above)
Comment (chinzou) [reply api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java:83]: Yes

File: api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java (hunk 18 above)
Comment (chinzou) [reply api/src/main/java/org/openmrs/module/initializer/api/form/BahmniFormsLoader.java:73]: Ok

File: readme/bahmniforms.md [hunk 23]
```diff
@@ -0,0 +1,43 @@
+## Domain 'bahmniforms'
+The **bahmniforms** subfolder contains [Bahmni Forms 2.0](https://bahmni.atlassian.net/wiki/spaces/BAH/pages/683966468/Forms+2.0) JSON definitions to provide the forms at runtime.
```
Comment (mks-d) [inline readme/bahmniforms.md:2]:
```
The **bahmniforms** subfolder contains [Bahmni Forms 2.0](https://bahmni.atlassian.net/wiki/spaces/BAH/pages/683966468/Forms+2.0) JSON definition files.
```
